app/models/miq_cockpit_ws_worker/runner.rb

```diff
@@ -0,0 +1,182 @@
+    end
+  end
+
+  def stop_cockpit_ws_process
+    return unless @pid
+    Process.kill("TERM", @pid)
+    wait_on_cockpit_ws
+  end
+
+  # Waits for a cockpit-ws process to stop.  The process is expected to be
+  #   in the act of shutting down, and thus it will wait 5 minutes
+  #   before issuing a kill.
+  def wait_on_cockpit_ws(pid)
+    pid ||= @pid
+    # TODO: Use Process.waitpid or one of its async variants
+    begin
+      Timeout.timeout(5.minutes.to_i) do
+        loop do
+          break unless process_alive?(pid)
+          sleep 1
+          heartbeat
+        end
+      end
+    rescue Timeout::Error
+      _log.info("#{log_prefix} Killing cockpit-ws process with pid=#{pid}")
+      Process.kill(9, pid)
+    end
+
+    begin
+      _, status = Process.waitpid2(pid)
+      _log.info("#{log_prefix} cockpit-ws Process with pid=#{pid} exited with a status=#{status}")
+    rescue Errno::ECHILD
+      _log.info("#{log_prefix} cockpit-ws Process with pid=#{pid} exited")
+    end
+
+    reset_process_info
+  end
+
+  def reset_process_info
+    @pid = @stdout = @stderr = nil
+  end
+
+  def find_cockpit_ws_processes
+    find_pids(/cockpit-ws --port/)
+  end
+
+  def stop_active_cockpit_ws_processes
+    find_cockpit_ws_processes.each do |pid|
+      _log.info("#{log_prefix} Killing active cockpit-ws process with pid=#{pid}")
+      Process.kill("TERM", pid)
+      wait_on_cockpit_ws(pid)
+    end
+  end
+
+  def cockpit_ws_alive?
+    if process_alive?(@pid)
+      true
+    else
+      reset_process_info
+      false
+    end
+  end
+
+  def process_alive?(pid)
+    return false if pid.nil?
+    process_struct = Sys::ProcTable.ps(pid)
+    return false if process_struct.nil?
+    return true if process_struct.state != "Z"
+
+    _log.info("#{log_prefix} waiting to die")
+    zombie_pid, status = Process.waitpid2(pid)
+    _log.info("#{log_prefix} cockpit-ws Process with pid=#{zombie_pid} exited with a status=#{status}")
+    false
+  end
+
+  def cockpit_ws_run
+    _log.info("#{log_prefix} cockpit-ws process starting")
+
+    opts = @worker_settings
+    cockpit_ws = MiqCockpit::WS.new(opts)
+    cockpit_ws.save_config
+
+    require 'open4'
+    env = {
+      "XDG_CONFIG_DIRS" => cockpit_ws.config_dir,
+      "DRB_URI"         => @drb_uri
+    }
+    pid, stdin, stdout, stderr = Open4.popen4(env, *cockpit_ws.command(BINDING_ADDRESS))
+    stdin.close
+
+    _log.info("#{log_prefix} cockpit-ws process started - pid=#{pid}")
+    return pid, stdout, stderr
+  end
+
+  def check_drb_service
+    alive = @drb_server ? @drb_server.alive? : false
+    unless alive
+      start_drb_service
+    end
+  end
+
+  def start_drb_service
+    require 'drb'
+    require 'drb/acl'
+
+    stop_drb_service
+    acl = ACL.new(%w(deny all allow 127.0.0.1/32))
+    @drb_server = DRb::DRbServer.new(@drb_uri || "druby://127.0.0.1:0",
+                                     MiqCockpitWsWorker::Authenticator.new, acl)
```

Comment:
Why are we sharing the authenticator through DRb?  Sorry, I don't really understand.